src/store/mqtt/actions.ts

```diff
@@ -39,81 +51,113 @@ export function connect() {
       password: jwt,
     });
 
-    dispatch(connecting());
-
-    handlers['connect'] = () => dispatch({ type: constants.CONNECTED });
-    handlers['close'] = () => dispatch(notConnected());
-    handlers['offline'] = () => dispatch(notConnected());
-    handlers['error'] = (err: any) => dispatch(setError(`Something went wrong: ${err}`));
-    handlers['message'] = (topic: string, message: MqttMessage) => dispatch(onMessage(topic, message));
-
-    // This is one of the most complex things in the code,
-    // the main purpose of this logic is to try to auto-reconnect after disconnecting.
-    // BUT, we cannot let mqtt.js handle this automatically for us, because our password
-    // (jwt) changes every few minutes (because that's why the session get's renewed and extended)
-    // so need to hijack the reconnection mechanism provided by mqtt.js, abort it (disconnect) and
-    // re initiate the connection ourselves, by dispatching(connect) recursively.
-    //
-    // Probably this whole thing would benefit from some refactor if you are up for it.
-    // TODO:
-    // - crate hooks so when the user logs out this is stopped
-    handlers['reconnect'] = async () => {
-      dispatch({ type: constants.RECONNECTING });
-      // save state.mqtt.data
-      const subscriptions = cloneDeep(getState().mqtt.data);
-      // disconnect
-      await dispatch(disconnect());
-      // connect
-      await dispatch(connect());
-      // re subscribe mqtt
-      client.subscribe(Object.keys(subscriptions), (error, granted) => {
-        if (error) {
-          dispatch(setError(error.message));
-          console.log(`MQTT: something went wrong while subscribing`, error);
+    try {
+      await new Promise((resolve, reject) => {
+        client.on('connect', resolve);
+        client.on('error', reject);
+      });
+    } catch (err) {
+      dispatch(setError(`Something went wrong: ${err}`));
+      dispatch({ type: constants.CONNECT_END });
+      throw err;
+    }
+
+    dispatch({ type: constants.CONNECTED });
+
+    client.on('error', (err: any) => dispatch(setError(`Parsing error: ${err}`)));
+    client.on('close', () => dispatch({ type: constants.CLOSE }));
+    client.on('offline', () => dispatch({ type: constants.OFFLINE }));
+    client.on('reconnect', () => dispatch(reconnect()));
+    client.on('message', (topic: string, message: MqttMessage) =>
```

Comment:
would also love to (at the application level) do what was noted in the issues section on MQTT.js's repo: ```In addition, we listen on AppState.addEventListener("change", etc...) and invoke mqttClient._checkPing() if the state is active in order to force the connection to attempt to reconnect when the app comes into focus because the JS timers don't work while it's in the background.```